prime-router/src/main/kotlin/fhirengine/azure/FHIRFunctions.kt

```diff
@@ -122,14 +127,29 @@ class FHIRFunctions(
         fhirEngine: FHIRTranslator,
         actionHistory: ActionHistory = ActionHistory(TaskAction.translate)
     ) {
-        val messageContent = readMessage("Translate", message, dequeueCount)
+        runFhirEngine(message, dequeueCount, fhirEngine, actionHistory)
+    }
 
-        try {
-            fhirEngine.doWork(messageContent, actionLogger, actionHistory)
-        } catch (e: Exception) {
-            logger.error("Unknown error.", e)
+    /**
+     * Deserializes the message, create the DB transaction and then runs the FHIR engine
+     *
+     * @return any messages that need to be dispatched
+     */
+    private fun runFhirEngine(
+        message: String,
+        dequeueCount: Int,
+        fhirEngine: FHIREngine,
+        actionHistory: ActionHistory,
+    ): List<RawSubmission> {
+        val messageContent = readMessage(fhirEngine.engineType, message, dequeueCount)
+
+        val newMessages = databaseAccess.transactReturning { txn ->
```

Comment:
So what happens if a runtime exception is thrown in `databaseAccess.transactReturning`?

Reply:
The transaction is not committed so no changes make it to the db